### PR TITLE
feat(email): Add flow events for email delivery notifications

### LINKED ISSUE #1528: Add an "email delivered" flow event
I just noticed that SES supports notifications for _successful_ delivery of email, along with notifications for bounces:

  https://aws.amazon.com/blogs/ses/announcing-delivery-notifications/

We should be able to use this to emit an "email delivered" flow event as another step in the sign-up/sign-in funnel.

### PR DESCRIPTION
**Pull request by vbudhram**
This PR adds the following to our email notifications
* Renamed `email_bouncer.js` to `email_notifications.js`
* Renamed `bounce` config property to `emailNotifications`
* Added support for processing email delivery notifications from SQS
* Logs flowEvents associated with delivery

~~While there are some code overlap in the `email-bounces.js` and `email-delivery.js` files, I thought it was more important to separate them because that seemed more logical to me?~~ Going to move common code to shared file but keep bounces and deliveries handling separate.

Fixes #1528

### CODE REVIEW

File: lib/email-delivery.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      // Ref: http://docs.aws.amazon.com/ses/latest/DeveloperGuide/notification-contents.html
+      const templateName = getHeaderValue('X-Template-Name', message)
+
+      return P.each(recipients, function (recipient) {
+
+        var email = recipient
+        var logData = {
+          op: 'handleDelivery',
+          email: email,
+          processingTimeMillis: message.delivery.processingTimeMillis
```
Comment (vbudhram) [inline lib/email-delivery.js]: Processing time was really the only noteworthy thing I thought was in the [delivery object](http://docs.aws.amazon.com/ses/latest/DeveloperGuide/notification-examples.html). Not against adding more

File: lib/email-delivery.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+var P = require('./promise')
+
+module.exports = function (log) {
+
+  return function start(deliveryQueue) {
+
+    function getHeaderValue(headerName, message){
```
Comment (philbooth) [inline lib/email-delivery.js]: Is this exactly the same as the one in the bouncer script? Should we extract it to a common module and `require` it in?

File: lib/email-delivery.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+          processingTimeMillis: message.delivery.processingTimeMillis
+        }
+
+        // Template name corresponds directly with the email template that was used
+        if (templateName) {
+          logData.template = templateName
+        }
+
+        // Log flow metrics if `flowId` and `flowBeginTime` specified in headers
+        const flowId = getHeaderValue('X-Flow-Id', message)
```
Comment (philbooth) [inline lib/email-delivery.js]: Looks like this stuff for emitting a flow event could go into a common function too right?

**Comment by rfk** [comment]
> Going to move common code to shared file but keep bounces and deliveries handling separate.

Worth checking with @jrgm and @jbuck from an ops perspective as well - would you prefer to set up delivery notifications to re-use the same SQS queue as the existing bounce/abuse events, or a separate queue?

**Comment by jbuck** [comment]
@rfk We current use separate queues for bounces and complaints, so I think we should use a separate queue for delivery notifications

File: lib/email-delivery.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+      const currentTime = Date.now()
+
+      var recipients = []
+      if (message.delivery && message.notificationType === 'Delivery') {
+        recipients = message.delivery.recipients
+      }
+
+      // SES can now send custom headers if enabled on topic.
+      // Headers are stored as an array of name/value pairs.
+      // Log the `X-Template-Name` header to help track the email template that bounced.
```
Comment (jrgm) [inline lib/email-delivery.js]: Fix comment. `s/bounced/delivered/`.

**Comment by vbudhram** [comment]
@philbooth Refactored based on comments, @mozilla/fxa-devs ?

**Comment by philbooth** [comment]
LGTM!

File: lib/email/bounces.js [hunk 5]
```diff
@@ -123,27 +105,8 @@ module.exports = function (log, error) {
...
-            time: currentTime,
-            flow_id: flowId,
-            flow_time: elapsedTime
-          }
-
-          log.flowEvent(flowEventInfo)
-        } else {
-          log.error({ op: 'handleBounce.flowEvent', templateName, flowId, flowBeginTime, currentTime })
-        }
+        // Log the bounced flowEvent metrics if avalible
```
Comment (vladikoff) [inline lib/email/bounces.js]: typo `avalible` -> `available`

**Comment by jrgm** [comment]
Excuse the drive-by comment, but why was this merged as 3 commits. Random fixups like 'PR fixes' and 'Fix typo' are not literate additions to the commit history.

**Comment by vbudhram** [comment]
@jrgm 😕, I used the github `squash and merge` command to merge this PR. Master only shows the merge commit and not the additional ones. Maybe I am misunderstanding how `github squash and merge` works?

**Pull request merged**